include/exadg/postprocessor/time_control.cpp

```diff
@@ -0,0 +1,213 @@
+    AssertThrow(
+      get_unsteady_evaluation_type(time_control_data_in) == TimeControlData::EvalType::Timestep,
+      dealii::ExcMessage("Sub timestep can only be used if TimeControlData::EvalType::Timestep."));
+
+  time_control_data = time_control_data_in;
+  counter           = time_control_data_in.counter_start;
+}
+
+// If setup() is not called needs_evaluation() will always evaluate to false
+bool
+TimeControl::needs_evaluation(double const time, types::time_step const time_step_number) const
+{
+  if(time_control_data.is_active == false)
+    return false;
+
+  // steady case
+  if(!Utilities::is_unsteady_timestep(time_step_number))
+  {
+    ++counter;
+    return true;
+  }
+
+  // unsteady evaluation
+  AssertThrow(get_unsteady_evaluation_type(time_control_data) != TimeControlData::EvalType::Invalid,
+              dealii::ExcMessage(
+                "Unsteady evaluation not possible since TimeControlData::EvalType::Invalid"));
+
+  if(time < time_control_data.start_time - EPSILON)
+    return false;
+
+  if(time > time_control_data.end_time - EPSILON)
+  {
+    if(time_control_data.force_final_evaluation)
+    {
+      if(forced_final_evaluation == false)
+      {
+        forced_final_evaluation = true;
+        ++counter;
+        return true;
+      }
+    }
+    else
+    {
+      return false;
+    }
+  }
+
+  bool evaluate = false;
+  if(get_unsteady_evaluation_type(time_control_data) == TimeControlData::EvalType::Timestep)
+  {
+    if((time_step_number - 1) % time_control_data.trigger_every_time_steps == 0)
+    {
+      evaluate = true;
+      ++counter;
+
+      if((time_step_number - 1) % (time_control_data.trigger_every_time_steps *
+                                   time_control_data.trigger_every_sub_time_step) ==
+         0)
+        is_sub_time_step_triggered = true;
+    }
+  }
+  else if(get_unsteady_evaluation_type(time_control_data) == TimeControlData::EvalType::Interval)
+  {
+    // In case of a restart, time might be significantly larger than start_time in the first time
+    // step of the restarted simulation. Hence, we have to reconstruct the counter at the end of the
+    // previous simulation, in order to obtain the correct behavior at the beginning of the
+    // restarted simulation.
+    if(reset_counter)
+    {
+      counter +=
+        int((time - time_control_data.start_time + EPSILON) / time_control_data.trigger_interval);
+      reset_counter = false;
+    }
+
+    if((time >
+        (time_control_data.start_time + counter * time_control_data.trigger_interval - EPSILON)))
+    {
+      evaluate = true;
+      ++counter;
+    }
+  }
+  else
+  {
+    AssertThrow(false, dealii::ExcMessage("Not implemented for given TimeControlData::EvalType"));
+  }
+
+
+
+  return evaluate;
+}
+
+unsigned int
+TimeControl::get_counter() const
+{
+  // the counter is incremented during needs_evaluation(), thus counter is ahead by 1
+  AssertThrow(counter - 1 != dealii::numbers::invalid_unsigned_int,
```

Comment:
is problematic if counter is of type unsigned. Did you think about this?

Reply:
Why is it problematic? Overflows are caught by the assert.

Reply:
what exactly is behind dealii::numbers::invalid_unsigned_int?

Reply:
static_cast<unsigned  int>(-1)

Reply:
I see. So this is a "complicated" way to write `counter > 0`?

Reply:
Again, I believe this implementation would have to take care of `counter_start` (but maybe I misunderstand the code?), a parameter which I am against introducing actually (also for such reasons).

Reply:
In this particular case, yes. I can change the assert.
I think this has nothing to do with counter_start. The counter is also invalid if the user forgot to call ```needs_evaluation()``` if the counter is 0 in the beginning. But this is OK, since we agreed that we can expect the user to call the function before any evaluation.